Tutorial ABCD method with RooParamtricHist

cesarecazzaniga opened the pull request:
Pull request to propose a small tutorial to illustrate the application of RooParametricHist for a per-bin ABCD method in Combine. 

Material added in the PR:
* documentation for the tutorial
* codes to generate input datasets, workspaces and cards to run the tutorial

kcormi reviewed (commented):
Thanks!

I've left some brief comments, I may have a few more later, but overall things look good. I am still debating whether it might be instructive to also e.g. set up the datacards with keywords, but than could always be added later.

Some of the 'code' comments were put in the md file where they are shown, but obviously apply to both places. I think we also need to update the mkdocs.yml so that this page would get included in the menu.

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+
+## Introduction
+The goal of this tutorial is to exemplify the usage of ```RooParametricHist``` in [CMS Combine](https://cms-analysis.github.io/HiggsAnalysis-CombinedLimit/latest/) to implement a bin-by-bin ABCD method.
+In this tutorial we will work with a toy example that could resamble a real physics analysis case. We consider the search for a BSM particle $\Phi$ with a mass range between 1500 and 5000 GeV that leads some excess in the tails of an observable $z$ (which could be $p_{T,\mathrm{miss}}$ ). We assume that we have found two uncorrelated discriminating features $x$ and $y$ that can be used to build t ...
````
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
```diff
@@ -0,0 +1,462 @@
+The tutorial has 4 main parts:
+
+1. [Generate input data](#inputs)
+2. [RooParametricHist for ABCD method](#rooparametrichist)
```
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+## Generate input data
+<a id="inputs"></a>
+
+The histograms for the $z$ observable in the different regions A,B,C,D can be produced using the ```produce_input_histograms_and_analyse.py``` in ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/data/tutorials/abcd_rooparametrichist_exercise/```. In the script the expected rates for different signal hypothesis (as a function of $\Phi$ mass $m_{\Phi} \in \{1500, 2000, 3000, 4000, 5000 \}$ GeV) and t ...
````
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+## RooParametricHist for ABCD method
+<a id="rooparametrichist"></a>
+
+In order to prepare the datacards for our ABCD method we will need to pass to the datacards ```shape``` section the histograms of our data in the A,B,C and D regions. 
````
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+In order to prepare the datacards for our ABCD method we will need to pass to the datacards ```shape``` section the histograms of our data in the A,B,C and D regions. 
+Moreover, we will need to relate the bins of our signal region A $N_{A}^{bin,i}$ to the bins of the control regions $N_{B/C/D}^{bin,i}$ via the ABCD method formula $N_{A}^{bin,i} = N_{B}^{bin,i} \cdot TF^{bin,i}$, where the transfer factor is $TF^{bin,i} = N_{C}^{bin,i}/N_{D}^{bin,i}$. To achieve our goal, we can use ```RooParametricHist``` implemented in Combine (for further documentation look [ ...
+
+A ```RooParametricHist``` is a custom implementation within the ROOT framework, specifically designed for handling parametric histograms in a way that integrates with ROOT's RooFit package. This class extends ```RooAbsPdf```, indicating it's meant to represent a probability density function (PDF) that can be parameterized and manipulated within the RooFit framework (see implentation in Combine [h ...
````
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+RooParametricHist parametric_hist("paramtric_hist", "Parametric Hist",variable,roo_arg_list_bins,data_th1)
+
+```
+where ```variable``` is a ```RooRealVar``` defining the observable we want to fit, ```roo_arg_list_bins``` is a ```RooArgList``` containing bins defined as ```RooRealVar``` or ```RooFormulaVar``` and ```data_th1``` is a ```TH1``` used to initialize the ```RooParametricHist```. We remark that it is also possible to define a normalization parameter for the parametric histogram as follows:
````
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+As an example, for each datacard, we have assigned a systematic uncertainty of 1.6% due to lumi to the signal processes, and a systematic of 5% to background in the SR (to take into account of non-closure of the method). 
+Notice that each datacard for each region has a ```shapes``` section for the observed data ```data_obs```, for the background ```Bkg``` and for the signal. The signal and data shapes are stored in a workspace ```wspace``` linked to the shapes section in the datacard, while the background shapes are stored in a ```RooParametricHist``` object. In the following we show how to build the workspace. 
+
+We follow the main steps implemented in a working code to create the workspace ```create_workspace.py``` in ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/data/tutorials/abcd_rooparametrichist_exercise/utils/```.
````
explanation moved above the part of datacards

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+```
+
+where ```-m``` is the flag for the mass point you want to run the script on. The script will use by default the histograms stored in ```generated_histograms```. To use the ones that you created, change the path in ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/data/tutorials/abcd_rooparametrichist_exercise/utils/create_datacards.py```.
+After running the script, the workspace will be saved in ```example_analysis/datacards/```. To create the datacards automatically fatching the correct workspace, run:
````
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
```diff
@@ -0,0 +1,462 @@
+
+#Add bins to RooArgList for the RooParametricHist in the B Region
+for idx,binB_i in enumerate(process_B_region_bins_list):
+    process_B_region_bins.add(binB_i)
```
I have removed the extra loop

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
```diff
@@ -0,0 +1,462 @@
+    #Compute the expected yield in the signal region as A = B * TF. This will be used to initialise the RooParametricHist in the Signal Region
+    bin_AB_i = RooFormulaVar("Bkg_AB_region_bin_"+str(i),"Background yield in SR A region bin " + str(i), "@0*@1", RooArgList(TF_i, ws.obj("Bkg_B_region_bin_"+str(i)) ))
+    process_AB_region_bins_list.append(bin_AB_i)
+for binAB_i in process_AB_region_bins_list:
```
I have removed the extra loop

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
````diff
@@ -0,0 +1,462 @@
+
+Both the observed (from nominal Monte Carlo) and the expected limits are computed for each mass point. 
+
+The same exercise can be repeated generating a workspace where the control regions are depleted from the signal (see datacards in ```$CMSSW_BASE/src/HiggsAnalysis/CombinedLimit/data/tutorials/abcd_rooparametrichist_exercise/datacards/no_sgn_CRs/```), and re-running the limits. This should give a hint of how much the signal contamination in the control regions is worsening the limits.  If you want ...
````
comments implemented

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
```diff
@@ -0,0 +1,462 @@
+
+<img src="figures/limits.png" width="600" />
+
+The analysis illustrated so far can be run for all mass point, and finally the typical brazil exclusion plot can be produced. As we can see the expected limit without signal in the control regions is better compared to the one taking into account the signal, showing that in this example there is an impact from the signal contamination of the control regions affecting the final sensitivity. Instea ...
```
i tried to stress that come from statistical fluctuations: "Instead, the observed line is showing the impact of statistical fluctuations inducing a non-closure effect of the ABCD method ...."

cesarecazzaniga commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
```diff
@@ -0,0 +1,462 @@
+
+The analysis illustrated so far can be run for all mass point, and finally the typical brazil exclusion plot can be produced. As we can see the expected limit without signal in the control regions is better compared to the one taking into account the signal, showing that in this example there is an impact from the signal contamination of the control regions affecting the final sensitivity. Instea ...
+
+In our single mass point analysis one can check indeed the behaviour illustrated in the limit plot above by looking at the $50$ % quantile on the signal strength $r$ after running the limit with and without signal injection in the control regions. Moreover, the effect of non-closure can be seen looking at the observed limit on $r$.
```
tried to specify as above

kcormi replied:
Thanks! I think the title is fine, at least I didn't manage to come up with a better one.

cesarecazzaniga commented:
Thanks for the comments and suggestions ! I went through them and updated the documentation and code. I left open three: one about the transfer factors (let me know if i got your point or you were meaning something different) and the two on non-closure. Finally, i have added the tutorial to the table of contents in mkdocs.yml as suggested, let me know if the title is fine.

kcormi commented on docs/abcd_rooparametrichist_tutorial/rooparametrichist_exercise.md:
```diff
@@ -0,0 +1,462 @@
+#Loop over the bins of the transfering region B, and compute the transfer factors as C/D
+for i in range(1,histB_pr.GetNbinsX()+1):
+    #Define transfer factor as a RooFormulaVar. Use the method .obj for RooWorkSpace to retrieve the yield for a given bin and region
+    TF_i = RooFormulaVar("TF"+str(i),"Transfer factor C/D bin " + str(i),"(@0/@1)",RooArgList(ws.obj("Bkg_C_region_bin_"+str(i)) , ws.obj("Bkg_D_region_bin_"+str(i)) ))
```
This part I certainly agree with! I think my comment wasnt very clear. I didn't mean not to use one per bin, I was just wondering about the use of the Transfer factors TF_i. 

They are an intermediate which is not really necessary, (but also completely fine to include!). The full equation is "A_i = B_i C_i / D_i" and this can be written in a number of different ways, including the way that you've done defining "TF_i = C_i / D_i". I am used to thinking of it in the first way, so I find it slightly clearer -- which I'm trying to consider in this pedagogical context, but I'm not sure if that's just me. Other people may find things clearer thinking with the TF_i intermediary.

cesarecazzaniga replied:
Now i see, thanks for the clarification ! As you prefer, if we talk about a "per-bin transfer factor method" then maybe might help to define TF_i in the code, but i agree in general is not necessary

nucleosynthesis commented:
Is it really necessary to include all of the results files (.root files from limits and log files etc) in the repository? Are these not things that the user will produce as they work through the exercises. There are many files being added here and it seems unecessary that we need to check them out each time we check out combine

cesarecazzaniga commented:
Yes, i can remove the .root files and just keep the datacards examples, i can even keep just 1 mass point example for the datacards
